File: JavaScript/ex/Deque.js
// addRear() 덱의 rear에 item을 추가하는 메서드
// addFront() 덱의 front에 item을 추가하는 메서드
// deleteRear() 덱의 rear에 있는 item을 반환한다.
// deleteFront() 덱의 front에 있는 item을 반환한다.
class DeckType {
  constructor(size) {
    this.maxQueueSize = size;
    this.array = [];
    this.front = 0;
    this.rear = 0;
  }

  isEmpty() {
    return this.front == this.rear;
  }

  isFull() {
    return (this.rear + 1) % this.maxQueueSize == this.front;
  }

  addFront(item) {
    if (this.isFull()) {
      console.log(new Error("큐가 포화상태입니다."));
    } else {
      this.array[this.front] = item;
      this.front = (this.front - 1 + this.maxQueueSize) % this.maxQueueSize;
    }
  }

  deleteFront() {
    if (this.isEmpty()) {
      console.log(new Error("큐가 비었습니다."));
    } else {
      const prev = this.front;
      this.front = (this.front + 1) % this.maxQueueSize;
      return this.array[prev];
    }
  }

  addRear(item) {
    if (this.isFull()) {
      console.log(new Error("큐가 포화상태입니다."));
    } else {
      this.rear = (this.rear + 1) % this.maxQueueSize;
      this.array[this.rear] = item;
    }
  }

  deleteRear() {
    if (this.isEmpty()) {
      console.log(new Error("큐가 비었습니다."));
    } else {
      const prev = this.rear;
      this.rear = (this.rear - 1 + this.maxQueueSize) % this.maxQueueSize;
      return this.array[prev];
    }
  }

  print() {
    if (this.isEmpty()) {
      console.log(new Error("큐가 비었습니다."));
    }
    let string = "";
    let i = this.front;
    do {
      i = (i + 1) % this.maxQueueSize;
      string += this.array[i] + "|";
      if (i == this.rear) {
        console.log(string);
        break;
      }
    } while (i != this.rear);
  }
}

File: Python/Deque.py
# add_front() 덱의 front에 item을 추가하는 메서드
# delete_front() 덱의 front에 있는 item을 반환한다.
# add_rear() 덱의 rear에 item을 추가하는 메서드
# delete_rear() 덱의 rear에 있는 item을 반환한다.

class DeckType:
  def __init__(self, size):
    self.max_queue_size = size
    self.array = [None] * size
    self.front = 0
    self.rear = 0

  def is_empty(self):
    return self.front == self.rear

  def is_full(self):
    return (self.rear + 1) % self.max_queue_size == self.front

  def add_front(self, item):
    if self.is_full():
      print("큐가 포화상태입니다.")
    else:
      self.array[self.front] = item
      self.front = (self.front - 1 + self.max_queue_size) % self.max_queue_size

  def delete_front(self):
    if self.is_empty():
      print("큐가 비었습니다.")
    else:
      prev = self.front
      self.front = (self.front + 1) % self.max_queue_size
      return self.array[prev]

  def add_rear(self, item):
    if self.is_full():
      print("큐가 포화상태입니다.")
    else:
      self.rear = (self.rear + 1) % self.max_queue_size
      self.array[self.rear] = item

  def delete_rear(self):
    if self.is_empty():
      print("큐가 비었습니다.")
    else:
      prev = self.rear
      self.rear = (self.rear - 1 + self.max_queue_size) % self.max_queue_size
      return self.array[prev]

  def print(self):
    if self.is_empty():
      print("큐가 비었습니다.")
    else:
      string = ""
      i = self.front
      while True:
        i = (i + 1) % self.max_queue_size
        string += str(self.array[i]) + "|"
        if i == self.rear:
          print(string)
          break
